lib/spack/spack/solver/concretize.lp

```diff
@@ -48,11 +32,27 @@ version_declared(Package, Version) :- version_declared(Package, Version, _).
 
 % If something is a package, it has only one version and that must be a
 % declared version.
-1 { version(Package, Version) : version_declared(Package, Version) } 1
- :- node(Package), error("Each node must have exactly one version").
-
-% A virtual package may have or not a version, but never has more than one
-:- virtual_node(Package), 2 { version(Package, _) }.
+% We allow clingo to choose any version(s), and infer an error if there
+% is not precisely one version chosen. Error facts are heavily optimized
+% against to ensure they cannot be inferred when a non-error solution is
+% possible
+{ version(Package, Version) : version_declared(Package, Version) }
+ :- node(Package).
+error(2, "No version for '{0}' satisfies '@{1}' and '@{2}'", Package, Version1, Version2)
```

Comment:
Minor: numbers in format strings aren't required unless you refer to something twice (e.g. in some of the error strings that construct specs).  We don't *have* to use numbers and can just use `{}`, in which case arguments are substituted in order.  This may make things easier to read where the number doesn't matter.

I wish we could write, e.g. `{Package}`, as `format()` supports named substitutions from kwargs, but I don't see how to easily do that on the Python side, as variable names aren't associated with the derived predicates, just the rule.  I think we'd need AST parsing for that, so probably best left as a nice-to-have for later.

Reply:
Does this basically boil down to whether we could reimplement python fstrings in older python versions for this?

Reply:
No -- I'm just saying you can do things like this even in 2.7:

```python
$ python
>>> print(
    "{Package} version {Version} is really complicated".format(
        Package="trilinos", Version="13.2.0",
    )
)
trilinos version 13.2.0 is really complicated
>>>
```

You don't need f-strings for it; you would just need to know the names of the variables in the rule on the ASP side.  But in `SpecBuilder`, you don't have them -- you'd have to parse the AST to figure out which positional args in the `error()` tuple correspond to which variable names from the rule that derived the it.